File: app/src/main/java/awais/instagrabber/managers/DirectMessagesManager.java
package awais.instagrabber.managers;

import android.content.ContentResolver;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.common.collect.Iterables;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.UUID;
import java.util.function.Function;

import awais.instagrabber.models.Resource;
import awais.instagrabber.repositories.requests.directmessages.BroadcastOptions;
import awais.instagrabber.repositories.responses.User;
import awais.instagrabber.repositories.responses.directmessages.DirectItem;
import awais.instagrabber.repositories.responses.directmessages.DirectThread;
import awais.instagrabber.repositories.responses.directmessages.DirectThreadBroadcastResponse;
import awais.instagrabber.repositories.responses.directmessages.RankedRecipient;
import awais.instagrabber.utils.Constants;
import awais.instagrabber.utils.CookieUtils;
import awais.instagrabber.webservices.DirectMessagesService;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static awais.instagrabber.utils.Utils.settingsHelper;

public final class DirectMessagesManager {
    private static final String TAG = DirectMessagesManager.class.getSimpleName();
    private static final Object LOCK = new Object();

    private static DirectMessagesManager instance;

    private final InboxManager inboxManager;
    private final InboxManager pendingInboxManager;

    private DirectMessagesService service;

    public static DirectMessagesManager getInstance() {
        if (instance == null) {
            synchronized (LOCK) {
                if (instance == null) {
                    instance = new DirectMessagesManager();
                }
            }
        }
        return instance;
    }

    private DirectMessagesManager() {
        inboxManager = InboxManager.getInstance(false);
        pendingInboxManager = InboxManager.getInstance(true);
        final String cookie = settingsHelper.getString(Constants.COOKIE);
        final long viewerId = CookieUtils.getUserIdFromCookie(cookie);
        final String deviceUuid = settingsHelper.getString(Constants.DEVICE_UUID);
        final String csrfToken = CookieUtils.getCsrfTokenFromCookie(cookie);
        if (csrfToken == null) return;
        service = DirectMessagesService.getInstance(csrfToken, viewerId, deviceUuid);
    }

    public void moveThreadFromPending(@NonNull final String threadId) {
        final List<DirectThread> pendingThreads = pendingInboxManager.getThreads().getValue();
        if (pendingThreads == null) return;
        final int index = Iterables.indexOf(pendingThreads, t -> t != null && t.getThreadId().equals(threadId));
        if (index < 0) return;
        final DirectThread thread = pendingThreads.get(index);
        final DirectItem threadFirstDirectItem = thread.getFirstDirectItem();
        if (threadFirstDirectItem == null) return;
        final List<DirectThread> threads = inboxManager.getThreads().getValue();
        int insertIndex = 0;
        if (threads != null) {
            for (final DirectThread tempThread : threads) {
                final DirectItem firstDirectItem = tempThread.getFirstDirectItem();
                if (firstDirectItem == null) continue;
                final long timestamp = firstDirectItem.getTimestamp();
                if (timestamp < threadFirstDirectItem.getTimestamp()) {
                    break;
                }
                insertIndex++;
            }
        }
        thread.setPending(false);
        inboxManager.addThread(thread, insertIndex);
        pendingInboxManager.removeThread(threadId);
        final Integer currentTotal = inboxManager.getPendingRequestsTotal().getValue();
        if (currentTotal == null) return;
        inboxManager.setPendingRequestsTotal(currentTotal - 1);
    }

    public InboxManager getInboxManager() {
        return inboxManager;
    }

    public InboxManager getPendingInboxManager() {
        return pendingInboxManager;
    }

    public ThreadManager getThreadManager(@NonNull final String threadId,
                                          final boolean pending,
                                          @NonNull final User currentUser,
                                          @NonNull final ContentResolver contentResolver) {
        return ThreadManager.getInstance(threadId, pending, currentUser, contentResolver);
    }

    public void createThread(final long userPk,
                             @Nullable final Function<DirectThread, Void> callback) {
        if (service == null) return;
        final Call<DirectThread> createThreadRequest = service.createThread(Collections.singletonList(userPk), null);
        createThreadRequest.enqueue(new Callback<DirectThread>() {
            @Override
            public void onResponse(@NonNull final Call<DirectThread> call, @NonNull final Response<DirectThread> response) {
                if (!response.isSuccessful()) {
                    if (response.errorBody() != null) {
                        try {
                            final String string = response.errorBody().string();
                            final String msg = String.format(Locale.US,
                                                             "onResponse: url: %s, responseCode: %d, errorBody: %s",
                                                             call.request().url().toString(),
                                                             response.code(),
                                                             string);
                            Log.e(TAG, msg);
                        } catch (IOException e) {
                            Log.e(TAG, "onResponse: ", e);
                        }
                        return;
                    }
                    Log.e(TAG, "onResponse: request was not successful and response error body was null");
                    return;
                }
                final DirectThread thread = response.body();
                if (thread == null) {
                    Log.e(TAG, "onResponse: thread is null");
                    return;
                }
                if (callback != null) {
                    callback.apply(thread);
                }
            }

            @Override
            public void onFailure(@NonNull final Call<DirectThread> call, @NonNull final Throwable t) {

            }
        });
    }

    public void sendMedia(@NonNull final Set<RankedRecipient> recipients, final String mediaId) {
        final int[] resultsCount = {0};
        final Function<Void, Void> callback = unused -> {
            resultsCount[0]++;
            if (resultsCount[0] == recipients.size()) {
                inboxManager.refresh();
            }
            return null;
        };
        for (final RankedRecipient recipient : recipients) {
            if (recipient == null) continue;
            sendMedia(recipient, mediaId, false, callback);
        }
    }

    public void sendMedia(@NonNull final RankedRecipient recipient, final String mediaId) {
        sendMedia(recipient, mediaId, true, null);
    }

    private void sendMedia(@NonNull final RankedRecipient recipient,
                           @NonNull final String mediaId,
                           final boolean refreshInbox,
                           @Nullable final Function<Void, Void> callback) {
        if (recipient.getThread() == null && recipient.getUser() != null) {
            // create thread and forward
            createThread(recipient.getUser().getPk(), directThread -> {
                sendMedia(directThread, mediaId, unused -> {
                    if (refreshInbox) {
                        inboxManager.refresh();
                    }
                    if (callback != null) {
                        callback.apply(null);
                    }
                    return null;
                });
                return null;
            });
        }
        if (recipient.getThread() == null) return;
        // just forward
        final DirectThread thread = recipient.getThread();
        sendMedia(thread, mediaId, unused -> {
            if (refreshInbox) {
                inboxManager.refresh();
            }
            if (callback != null) {
                callback.apply(null);
            }
            return null;
        });
    }

    @NonNull
    public LiveData<Resource<Object>> sendMedia(@NonNull final DirectThread thread,
                                                @NonNull final String mediaId,
                                                @Nullable final Function<Void, Void> callback) {
        return sendMedia(thread.getThreadId(), mediaId, callback);
    }

    @NonNull
    public LiveData<Resource<Object>> sendMedia(@NonNull final String threadId,
                                                @NonNull final String mediaId,
                                                @Nullable final Function<Void, Void> callback) {
        final MutableLiveData<Resource<Object>> data = new MutableLiveData<>();
        data.postValue(Resource.loading(null));
        final Call<DirectThreadBroadcastResponse> request = service.broadcastMediaShare(
                UUID.randomUUID().toString(),
                BroadcastOptions.ThreadIdOrUserIds.of(threadId),
                mediaId
        );
        request.enqueue(new Callback<DirectThreadBroadcastResponse>() {
            @Override
            public void onResponse(@NonNull final Call<DirectThreadBroadcastResponse> call,
                                   @NonNull final Response<DirectThreadBroadcastResponse> response) {
                if (response.isSuccessful()) {
                    data.postValue(Resource.success(new Object()));
                    if (callback != null) {
                        callback.apply(null);
                    }
                    return;
                }
                if (response.errorBody() != null) {
                    try {
                        final String string = response.errorBody().string();
                        final String msg = String.format(Locale.US,
                                                         "onResponse: url: %s, responseCode: %d, errorBody: %s",
                                                         call.request().url().toString(),
                                                         response.code(),
                                                         string);
                        Log.e(TAG, msg);
                        data.postValue(Resource.error(msg, null));
                    } catch (IOException e) {
                        Log.e(TAG, "onResponse: ", e);
                        data.postValue(Resource.error(e.getMessage(), null));
                    }
                    if (callback != null) {
                        callback.apply(null);
                    }
                    return;
                }
                final String msg = "onResponse: request was not successful and response error body was null";
                Log.e(TAG, msg);
                data.postValue(Resource.error(msg, null));
                if (callback != null) {
                    callback.apply(null);
                }
            }

            @Override
            public void onFailure(@NonNull final Call<DirectThreadBroadcastResponse> call, @NonNull final Throwable t) {
                Log.e(TAG, "onFailure: ", t);
                data.postValue(Resource.error(t.getMessage(), null));
                if (callback != null) {
                    callback.apply(null);
                }
            }
        });
        return data;
    }
}

File: app/src/main/java/awais/instagrabber/managers/DirectMessagesManager.kt
package awais.instagrabber.managers

import android.content.ContentResolver
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import awais.instagrabber.models.Resource
import awais.instagrabber.models.Resource.Companion.error
import awais.instagrabber.models.Resource.Companion.loading
import awais.instagrabber.models.Resource.Companion.success
import awais.instagrabber.models.enums.BroadcastItemType
import awais.instagrabber.repositories.requests.directmessages.ThreadIdsOrUserIds
import awais.instagrabber.repositories.responses.User
import awais.instagrabber.repositories.responses.directmessages.DirectThread
import awais.instagrabber.repositories.responses.directmessages.RankedRecipient
import awais.instagrabber.utils.Constants
import awais.instagrabber.utils.Utils
import awais.instagrabber.utils.getCsrfTokenFromCookie
import awais.instagrabber.utils.getUserIdFromCookie
import awais.instagrabber.webservices.DirectMessagesRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*

object DirectMessagesManager {
    val inboxManager: InboxManager by lazy { InboxManager(false) }
    val pendingInboxManager: InboxManager by lazy { InboxManager(true) }

    private val TAG = DirectMessagesManager::class.java.simpleName
    private val viewerId: Long
    private val deviceUuid: String
    private val csrfToken: String
    private val directMessagesRepository by lazy { DirectMessagesRepository.getInstance() }

    fun moveThreadFromPending(threadId: String) {
        val pendingThreads = pendingInboxManager.threads.value ?: return
        val index = pendingThreads.indexOfFirst { it.threadId == threadId }
        if (index < 0) return
        val thread = pendingThreads[index]
        val threadFirstDirectItem = thread.firstDirectItem ?: return
        val threads = inboxManager.threads.value
        var insertIndex = 0
        if (threads != null) {
            for (tempThread in threads) {
                val firstDirectItem = tempThread.firstDirectItem ?: continue
                val timestamp = firstDirectItem.getTimestamp()
                if (timestamp < threadFirstDirectItem.getTimestamp()) {
                    break
                }
                insertIndex++
            }
        }
        thread.pending = false
        inboxManager.addThread(thread, insertIndex)
        pendingInboxManager.removeThread(threadId)
        val currentTotal = inboxManager.getPendingRequestsTotal().value ?: return
        inboxManager.setPendingRequestsTotal(currentTotal - 1)
    }

    fun getThreadManager(
        threadId: String,
        pending: Boolean,
        currentUser: User,
        contentResolver: ContentResolver,
    ): ThreadManager {
        return ThreadManager(threadId, pending, currentUser, contentResolver, viewerId, csrfToken, deviceUuid)
    }

    suspend fun createThread(userPk: Long): DirectThread =
        directMessagesRepository.createThread(csrfToken, viewerId, deviceUuid, listOf(userPk), null)

    fun sendMedia(recipient: RankedRecipient, mediaId: String, secondId: String?, itemType: BroadcastItemType, scope: CoroutineScope) {
        sendMedia(setOf(recipient), mediaId, secondId, itemType, scope)
    }

    fun sendMedia(
        recipients: Set<RankedRecipient>,
        mediaId: String,
        secondId: String?,
        itemType: BroadcastItemType,
        scope: CoroutineScope,
    ) {
        val threadIds = recipients.mapNotNull { it.thread?.threadId }
        val userIdsTemp = recipients.mapNotNull { it.user?.pk }
        val userIds = userIdsTemp.map { listOf(it.toString(10)) }
        sendMedia(threadIds, userIds, mediaId, secondId, itemType, scope) {
            inboxManager.refresh(scope)
        }
    }

    private fun sendMedia(
        threadIds: List<String>,
        userIds: List<List<String>>,
        mediaId: String,
        secondId: String?,
        itemType: BroadcastItemType,
        scope: CoroutineScope,
        callback: (() -> Unit)?,
    ): LiveData<Resource<Any?>> {
        val data = MutableLiveData<Resource<Any?>>()
        data.postValue(loading(null))
        scope.launch(Dispatchers.IO) {
            try {
                when (itemType) {
                    BroadcastItemType.MEDIA_SHARE -> directMessagesRepository.broadcastMediaShare(
                        csrfToken,
                        viewerId,
                        deviceUuid,
                        UUID.randomUUID().toString(),
                        ThreadIdsOrUserIds(threadIds, userIds),
                        mediaId,
                        secondId
                    )
                    BroadcastItemType.PROFILE -> directMessagesRepository.broadcastProfile(
                        csrfToken,
                        viewerId,
                        deviceUuid,
                        UUID.randomUUID().toString(),
                        ThreadIdsOrUserIds(threadIds, userIds),
                        mediaId
                    )
                    BroadcastItemType.STORY -> directMessagesRepository.broadcastStory(
                        csrfToken,
                        viewerId,
                        deviceUuid,
                        UUID.randomUUID().toString(),
                        ThreadIdsOrUserIds(threadIds, userIds),
                        mediaId,
                        secondId!!
                    )
                }
                data.postValue(success(Any()))
                callback?.invoke()
            } catch (e: Exception) {
                Log.e(TAG, "sendMedia: ", e)
                data.postValue(error(e.message, null))
                callback?.invoke()
            }
        }
        return data
    }

    fun replyToStory(
        recipientId: Long?,
        reelId: String?,
        mediaId: String?,
        text: String,
        scope: CoroutineScope
    ): LiveData<Resource<Any?>> {
        Log.d("austin_debug", "replying")
        val data = MutableLiveData<Resource<Any?>>()
        data.postValue(loading(null))
        if (recipientId == null || reelId == null || mediaId == null) {
            data.postValue(error("arguments are null", null))
            return data
        }
        scope.launch(Dispatchers.IO) {
            try {
                directMessagesRepository.broadcastStoryReply(
                    csrfToken,
                    viewerId,
                    deviceUuid,
                    ThreadIdsOrUserIds.Companion.ofOneUser(recipientId.toString(10)),
                    text,
                    mediaId,
                    reelId
                )
                inboxManager.refresh(scope)
                data.postValue(success(null))
            }
            catch (e: Exception) {
                Log.e(TAG, "story reply: ", e)
                data.postValue(error(e.message, null))
            }
        }
        return data
    }

    init {
        val cookie = Utils.settingsHelper.getString(Constants.COOKIE)
        viewerId = getUserIdFromCookie(cookie)
        deviceUuid = Utils.settingsHelper.getString(Constants.DEVICE_UUID)
        val csrfToken = getCsrfTokenFromCookie(cookie)
        require(!csrfToken.isNullOrBlank() && viewerId != 0L && deviceUuid.isNotBlank()) { "User is not logged in!" }
        this.csrfToken = csrfToken
    }
}